src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -8120,15 +8132,20 @@ void checkWellKnown(WellKnownMember member)
             }
         }
 
-        private bool TryFindIndexOrRangeIndexerPattern(
-            BoundExpression? receiverOpt,
-            TypeSymbol receiverType,
+        /// <summary>
+        /// Finds pattern-based implicit indexer and Length/Count property.
+        /// We'll use the receiver placeholder to bind the length access, but the real
+        /// receiver to bind the indexer access.
```

Comment:
This feels odd to me. I haven't yet looked at lowering, but it strikes me that this means that, to replace the placeholder for the count, we'll have to look into the indexer, find the the real thing, evaluate it to a temp, then add a placeholder replacement. It feels like this would be cleaner using a placeholder in both places.

Reply:
What you're describing is what I'd done earlier in this PR. The issue is that CheckValue doesn't handle placeholders. By keeping the receiver in IndexerAccess, CheckValue can at least correctly check that one.
Pulling the Receiver out of IndexerAccess is tracked by https://github.com/dotnet/roslyn/issues/57855

Reply:
Thanks for the clarification, that makes sense.